File: app/src/main/java/de/eloc/eloc_control_panel/activities/MapActivity.java
package de.eloc.eloc_control_panel.activities;

import android.annotation.SuppressLint;
import android.location.Location;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.Marker;
import com.google.maps.android.clustering.ClusterManager;
import com.google.openlocationcode.OpenLocationCode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Locale;

import de.eloc.eloc_control_panel.R;
import de.eloc.eloc_control_panel.databinding.ActivityMapBinding;
import de.eloc.eloc_control_panel.databinding.WindowLayoutBinding;
import de.eloc.eloc_control_panel.helpers.HttpHelper;
import de.eloc.eloc_control_panel.models.ElocDeviceInfo;
import de.eloc.eloc_control_panel.models.ElocMarker;

public class MapActivity extends AppCompatActivity {
    /*
    For reference, use the info below to help set default zoom level for when a marker is tapped:
    1: World
    5: Landmass/continent
    10: City
    15: Streets
    20: Buildings
     */
    private static final int MARKER_ZOOM = 12;
    private static final int INITIAL_ZOOM = 10; // City level
    private int customZoom = 2;
    private ArrayList<String> unknowDevices = new ArrayList<>();
    private ActivityMapBinding binding;
    private GoogleMap map = null;
    private LatLngBounds mapBounds = null;
    private ArrayList<ElocDeviceInfo> devices = null;
    private ClusterManager<ElocMarker> clusterManager = null;
    private final HashSet<String> usedLocations = new HashSet<>();
    private final GoogleMap.InfoWindowAdapter infoWindowAdapter = new GoogleMap.InfoWindowAdapter() {

        @Nullable
        @Override
        public View getInfoContents(@NonNull Marker marker) {
            return null;
        }

        @Override
        public View getInfoWindow(@NonNull Marker marker) {
            // When a marker info window is shown, reset lastCustomZoom
            // so that when clusters are clicked on later, the map will not zoom too deep
            customZoom = INITIAL_ZOOM;

            zoomTo(MARKER_ZOOM, marker.getPosition());
            WindowLayoutBinding windowLayoutBinding = WindowLayoutBinding.inflate(getLayoutInflater());
            windowLayoutBinding.titleTextView.setText(marker.getTitle());
            windowLayoutBinding.snippetTextView.setText(marker.getSnippet());
            return windowLayoutBinding.getRoot();
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMapBinding.inflate(getLayoutInflater());
        binding.mapView.setVisibility(View.INVISIBLE);

        setContentView(binding.getRoot());
        setToolbar();
        setListeners();
        hideUnknownDevices();

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(binding.mapView.getId());
        if (mapFragment != null) {
            mapFragment.getMapAsync(googleMap -> {
                map = googleMap;
                showMap();
            });
        }
        HttpHelper.getElocDevicesAsync(this::elocDeviceInfoReceived);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return false;
    }

    private void setToolbar() {
        binding.appbar.toolbar.setTitle(R.string.find_your_elocs);
        setSupportActionBar(binding.appbar.toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    private void elocDeviceInfoReceived(ArrayList<ElocDeviceInfo> infos) {
        devices = infos;
        showMap();
    }

    private void zoomTo(int zoomLevel, LatLng position) {
        // Set a delay so that cluster manager does not interfere with zoom animation
        int delayMillis = 1200;
        Handler handler = new Handler(Looper.getMainLooper());
        handler.postDelayed(
                () -> map.animateCamera(CameraUpdateFactory.newCameraPosition(CameraPosition.builder().zoom(zoomLevel).target(position).build()))
                , delayMillis
        );

    }

    private void setListeners() {
        binding.unknownDevicesButton.setOnClickListener(v -> {
            if (binding.upIcon.getVisibility() == View.VISIBLE) {
                collapseUnknownDevices();
            } else {
                expandUnknownDevices();
            }
        });
    }

    private void expandUnknownDevices() {
        binding.upIcon.setVisibility(View.VISIBLE);
        binding.downIcon.setVisibility(View.GONE);
        binding.unknownDevicesTextView.setVisibility(View.VISIBLE);
    }

    private void collapseUnknownDevices() {
        binding.upIcon.setVisibility(View.GONE);
        binding.downIcon.setVisibility(View.VISIBLE);
        binding.unknownDevicesTextView.setVisibility(View.GONE);
    }

    private void hideUnknownDevices() {
        binding.unknownDevicesPanel.setVisibility(View.GONE);
    }

    private void showUnknownDevices() {
        binding.unknownDevicesPanel.setVisibility(View.VISIBLE);
    }

    @SuppressLint("MissingPermission")
    private void showMap() {
        if (map == null || devices == null) {
            return;
        }

        runOnUiThread(() -> {
            // Note: No need for permission check here
            // because PermissionsActivity ensures that app will only launch
            // when all required permission are granted, including location.

            map.setMyLocationEnabled(true);
            map.getUiSettings().setMapToolbarEnabled(false);
            binding.mapView.setVisibility(View.VISIBLE);
            binding.loadingLayout.setVisibility(View.GONE);

            clusterManager = new ClusterManager<>(this, map);
            clusterManager.getMarkerCollection().setInfoWindowAdapter(infoWindowAdapter);
            clusterManager.setOnClusterClickListener(cluster -> {
                // Changing the zoom level on each tap so that map keeps zooming in.
                if (customZoom < INITIAL_ZOOM) {
                    customZoom = INITIAL_ZOOM;
                } else {
                    customZoom += 5;
                }
                zoomTo(customZoom, cluster.getPosition());
                return true;
            });
            map.setOnCameraIdleListener(clusterManager);
            showDevices();
        });
    }


    private void showDevices() {
        mapBounds = null;
        if (clusterManager != null) {
            clusterManager.clearItems();
            usedLocations.clear();
        }
        unknowDevices.clear();
        for (ElocDeviceInfo device : devices) {
            addMarker(device);
        }
        if (clusterManager != null) {
            clusterManager.cluster();
        }
    }

    private void updateUnknownDevices() {
        StringBuilder builder = new StringBuilder();
        String[] names = unknowDevices.toArray(new String[]{});
        Arrays.sort(names);
        int maxIndex = names.length - 1;
        for (int i = 0; i <= maxIndex; i++) {
            builder.append(names[i]);
            if (i < maxIndex) {
                builder.append(", ");
            }
        }
        binding.unknownDevicesTextView.setText(builder.toString());
    }

    private void addMarker(ElocDeviceInfo device) {
        OpenLocationCode.CodeArea decodedVal = null;
        try {
            decodedVal = OpenLocationCode.decode(device.plusCode);
        } catch (IllegalArgumentException ignore) {

        }

        // If location code is invalid, add to unknown list and skip
        if (decodedVal == null) {
            unknowDevices.add(device.name);
            showUnknownDevices();
            expandUnknownDevices();
            updateUnknownDevices();
            return;
        }

        LatLng location = new LatLng(decodedVal.getCenterLatitude(), decodedVal.getCenterLongitude());
        String stringifiedLocation = location.latitude + ":" + location.longitude;
        double offset = -1;
        if (usedLocations.contains(stringifiedLocation)) {
            LatLng offsetLocation = new LatLng(location.latitude + 0.00005, location.longitude);
            float[] distance = new float[1];
            Location.distanceBetween(location.latitude, location.longitude, offsetLocation.latitude, offsetLocation.longitude, distance);
            location = offsetLocation;
            offset = distance[0];
        }
        if (mapBounds == null) {
            mapBounds = new LatLngBounds(location, location);
        } else {
            mapBounds = mapBounds.including(location);
        }
        map.setLatLngBoundsForCameraTarget(mapBounds);

        if (clusterManager != null) {
            String snippet = "Last Update: " + device.time + "\n" +
                    "\n" +
                    "GPS Accuracy: " + getPretty(device.accuracy) + "\n" +
                    "\n" +
                    "Battery volts: " + getPretty(device.batteryVolts) + "\n" +
                    "\n" +
                    "Rec Time: " + getPretty(device.recTime);
            if (offset > 0) {
                snippet += String.format(Locale.ENGLISH, "\n\n(Offset by ~ %.2fm)", offset);
            }
            boolean added = clusterManager.addItem(new ElocMarker(device.name, snippet, location));
            if (added) {
                usedLocations.add(stringifiedLocation);
            }
        }
    }

    private String getPretty(double value) {
        String formatString = "%.2f";
        return String.format(Locale.ENGLISH, formatString, value);
    }
}

File: app/src/main/java/de/eloc/eloc_control_panel/activities/MapActivity.kt
package de.eloc.eloc_control_panel.activities

import android.annotation.SuppressLint
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.View
import android.widget.ArrayAdapter
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.InfoWindowAdapter
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.Marker
import com.google.maps.android.clustering.ClusterManager
import de.eloc.eloc_control_panel.R
import de.eloc.eloc_control_panel.data.AppState
import de.eloc.eloc_control_panel.data.ElocDeviceInfo
import de.eloc.eloc_control_panel.data.ElocMarker
import de.eloc.eloc_control_panel.data.adapters.MapInfoAdapter
import de.eloc.eloc_control_panel.data.helpers.LocationHelper
import de.eloc.eloc_control_panel.data.helpers.firebase.FirestoreHelper
import de.eloc.eloc_control_panel.databinding.ActivityMapBinding
import de.eloc.eloc_control_panel.databinding.WindowLayoutBinding
import de.eloc.eloc_control_panel.dialogs.ListViewDialog
import java.util.Locale

class MapActivity : ThemableActivity() {

    private lateinit var binding: ActivityMapBinding
    private val unknownDevices = mutableSetOf<String>()
    private val usedLocations = mutableSetOf<String>()
    private var map: GoogleMap? = null
    private var mapDevices = mutableMapOf<String, ElocDeviceInfo>()
    private var deviceCache = mutableListOf<ElocDeviceInfo>()
    private var mapBounds: LatLngBounds? = null
    private var clusterManager: ClusterManager<ElocMarker>? = null

    private val infoWindowAdapter: InfoWindowAdapter = object : InfoWindowAdapter {
        override fun getInfoContents(marker: Marker): View? {
            return null
        }

        override fun getInfoWindow(marker: Marker): View {
            // When a marker info window is shown, reset lastCustomZoom
            // so that when clusters are clicked on later, the map will not zoom too deep
            val windowLayoutBinding: WindowLayoutBinding =
                WindowLayoutBinding.inflate(layoutInflater)
            windowLayoutBinding.titleTextView.text = marker.title
            windowLayoutBinding.snippetTextView.text = marker.snippet
            return windowLayoutBinding.root
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMapBinding.inflate(layoutInflater)
        binding.mapView.visibility = View.INVISIBLE
        binding.progressIndicator.visibility = View.VISIBLE
        binding.dialogBackground.visibility = View.GONE

        setContentView(binding.root)
        menuInflater.inflate(R.menu.find_elocs_menu, binding.toolbar.menu)
        setListeners()

        if (AppState.hasValidProfile) {
            val mapFragment =
                supportFragmentManager.findFragmentById(binding.mapView.id) as SupportMapFragment?
            mapFragment?.getMapAsync { googleMap: GoogleMap ->
                map = googleMap
                initializeMap()
            }

            FirestoreHelper.instance.getElocDevicesAsync { info ->
                runOnUiThread {
                    deviceCache.add(info)
                    if (map != null) {
                        showMarkers()
                    }
                }
            }
        } else {
            showModalAlert(
                getString(R.string.required),
                getString(R.string.ranger_name_required)
            ) {
                goBack()
            }
        }
    }

    private fun setListeners() {
        binding.toolbar.setNavigationOnClickListener { goBack() }
        binding.toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.mnu_unknown_location -> showElocsWithUnknownLocation()
                R.id.mnu_list_view -> showElocList()
            }
            true
        }
    }

    private fun zoomIn(position: LatLng) {
        // Set a delay so that cluster manager does not interfere with zoom animation
        val delayMillis = 800
        val handler = Handler(Looper.getMainLooper())
        handler.postDelayed(
            {
                if (map == null) {
                    return@postDelayed
                }

                /*
                For reference, use the info below to help set default zoom level for when a marker is tapped:
                1: World
                5: Landmass/continent
                10: City
                15: Streets
                20: Buildings
                */

                val cameraPosition = CameraPosition.builder()
                    .zoom(map!!.cameraPosition.zoom + 3)
                    .target(position)
                    .build()
                map?.animateCamera(CameraUpdateFactory.newCameraPosition(cameraPosition))
            }, delayMillis
                .toLong()
        )
    }

    private fun initializeMap() {
        synchronized(map!!) {
            runOnUiThread {
                if (map == null) {
                    return@runOnUiThread
                }

                // Note: No need for permission check here
                // because PermissionsActivity ensures that app will only launch
                // when all required permission are granted, including location.
                @SuppressLint("MissingPermission")
                map?.isMyLocationEnabled = true
                binding.mapView.visibility = View.VISIBLE
                binding.progressIndicator.visibility = View.GONE
                clusterManager = ClusterManager(this, map)

                clusterManager?.markerCollection?.setInfoWindowAdapter(infoWindowAdapter)
                clusterManager?.setOnClusterClickListener { cluster ->
                    // Changing the zoom level on each tap so that map keeps zooming in.
                    zoomIn(cluster.position)
                    true
                }
                map?.setOnCameraIdleListener(clusterManager)
                showMarkers()
            }
        }
    }


    private fun showMarkers() {
        var infoAdded = false
        while (deviceCache.isNotEmpty()) {
            val newInfo = deviceCache.removeAt(0)
            // Add to map (replace older)
            val key = newInfo.name
            var add = true
            if (mapDevices.containsKey(key)) {
                val oldInfo = mapDevices[key]
                if (oldInfo != null) {
                    if (oldInfo.time.compareTo(newInfo.time) >= 1) {
                        add = false
                    }
                }
            }

            if (add) {
                infoAdded = true
                mapDevices[key] = newInfo
            }
        }

        if (infoAdded) {
            mapBounds = null
            clusterManager?.clearItems()
            usedLocations.clear()
            unknownDevices.clear()
            for (info in mapDevices) {
                addMarker(info.value)
            }
            clusterManager?.cluster()
        }
    }

    private fun getPretty(value: Double): String {
        return getString(R.string.pretty_double_template, value)
    }

    private fun addMarker(device: ElocDeviceInfo) {
        val loc = device.location
        if (loc == null) {
            unknownDevices.add(device.name)
        } else {
            var location = loc
            var stringifiedLocation = LocationHelper.prettifyLocation(location)
            var offset: Double = -1.0
            while (usedLocations.contains(stringifiedLocation)) {
                val offsetLocation =
                    LatLng(location!!.latitude + 0.00005, location.longitude)
                val distance = FloatArray(1)
                Location.distanceBetween(
                    location.latitude,
                    location.longitude,
                    offsetLocation.latitude,
                    offsetLocation.longitude,
                    distance
                )
                location = offsetLocation
                stringifiedLocation = LocationHelper.prettifyLocation(location)
                offset = distance[0].toDouble()
            }
            mapBounds = if (mapBounds == null) {
                LatLngBounds(location!!, location)
            } else {
                mapBounds!!.including(location!!)
            }

            // Move camera to center of map bounds, but do not set bounds on map/camera
            //map?.setLatLngBoundsForCameraTarget(mapBounds)
            val centerOfMapBounds = CameraPosition.builder()
                .target(mapBounds!!.center)
                .zoom(8f)
                .build()
            map?.animateCamera(CameraUpdateFactory.newCameraPosition(centerOfMapBounds))

            if (clusterManager != null) {
                var snippet = getString(
                    R.string.snippet_template,
                    device.time,
                    getPretty(device.accuracy),
                    getPretty(device.batteryVolts),
                    getPretty(device.recTime)
                )
                if (offset > 0) {
                    snippet += String.format(Locale.ENGLISH, "\n\n(Offset by ~ %.2fm)", offset)
                }
                val added =
                    clusterManager?.addItem(ElocMarker(device.name, snippet, location)) ?: false
                if (added) {
                    usedLocations.add(stringifiedLocation)
                }
            }
        }
    }

    @SuppressLint("InternalInsetResource", "DiscouragedApi")
    private fun getStatusBarHeight(): Int {
        val resId = resources.getIdentifier("status_bar_height", "dimen", "android")
        return if (resId > 0) {
            resources.getDimensionPixelSize(resId)
        } else {
            (resources.displayMetrics.density * 25).toInt()
        }
    }

    private fun getActionBarSize(): Int {
        val tv = TypedValue()
        return if (theme.resolveAttribute(android.R.attr.actionBarSize, tv, true)) {
            TypedValue.complexToDimensionPixelSize(tv.data, resources.displayMetrics)
        } else {
            (resources.displayMetrics.heightPixels * 36)
        }
    }

    private fun calculateDialogHeight(): Int {
        val statusBarHeight = getStatusBarHeight()
        val screenHeight = resources.displayMetrics.heightPixels
        return screenHeight - getActionBarSize() - statusBarHeight //- statusBarHeight
    }

    private fun showElocsWithUnknownLocation() {
        val sortedItems = if (unknownDevices.isEmpty()) {
            listOf(getString(R.string.none))
        } else {
            unknownDevices.sorted()
        }
        val adapter =
            ArrayAdapter(this, android.R.layout.simple_list_item_1, sortedItems)
        val height = calculateDialogHeight()
        val dialog = ListViewDialog(
            getString(R.string.elocs_with_unknown_location),
            height,
            adapter,
            { binding.dialogBackground.visibility = View.VISIBLE },
            { binding.dialogBackground.visibility = View.GONE }
        )
        dialog.show(supportFragmentManager)
    }

    private fun showElocList() {
        val adapter = if (mapDevices.isEmpty()) {
            ArrayAdapter(
                this,
                android.R.layout.simple_list_item_1,
                listOf(getString(R.string.none))
            )
        } else {
            val sortedItems = mapDevices.values.toList().sortedWith { a, b ->
                a.name.compareTo(b.name)
            }
            MapInfoAdapter(this, R.layout.layout_map_table_item, sortedItems)
        }
        val height = calculateDialogHeight()
        val dialog = ListViewDialog(
            getString(R.string.my_elocs),
            height,
            adapter,
            { binding.dialogBackground.visibility = View.VISIBLE },
            { binding.dialogBackground.visibility = View.GONE }
        )
        dialog.show(supportFragmentManager)
    }
}
